### ISSUE TITLE
False positive in Go module on a dependency that is overridden with a safe version at compile time.

### ISSUE THREAD

**Issue by jeffwidman** [aquasecurity/trivy#1559]
I compile/install [`grpcurl`](https://github.com/fullstorydev/grpcurl) `v1.8.5` into an image by having a dummy `main.go` file:
```
package main

import (
	"fmt"

	_ "github.com/fullstorydev/grpcurl"
)

func main() {
	fmt.Println(`
This is a dummy application used to pull in a go.mod dependency with grpcurl
which allows us to trigger a dependabot upgrade automatically of the grpcurl
tool.`)
}
```

I tried running Trivy against the resulting image and got the following warning:
```
991usr/local/bin/grpcurl (gobinary)
992================================
993Total: 1 (HIGH: 1, CRITICAL: 0)
994
995+-------------------+------------------+----------+-------------------+---------------+---------------------------------------+
996|      LIBRARY      | VULNERABILITY ID | SEVERITY | INSTALLED VERSION | FIXED VERSION |                 TITLE                 |
997+-------------------+------------------+----------+-------------------+---------------+---------------------------------------+
998| golang.org/x/text | CVE-2020-14040   | HIGH     | v0.3.2            | 0.3.3         | golang.org/x/text: possibility        |
999|                   |                  |          |                   |               | to trigger an infinite loop in        |
1000|                   |                  |          |                   |               | encoding/unicode could lead to...     |
1001|                   |                  |          |                   |               | -->avd.aquasec.com/nvd/cve-2020-14040 |
1002+-------------------+------------------+----------+-------------------+---------------+---------------------------------------+
```

`golang.org/x/text` is an indirect dependency. So I bumped the version number in the `go.mod` file using the following pin:

```
module github.com/<redacted>

go 1.16

require (
	github.com/fullstorydev/grpcurl v1.8.5
	golang.org/x/text v0.3.3 // indirect; required to fix CVE-2020-14040
)
```

This will cause the compiled version to only include `v0.3.3` which fixes the CVE. 

Unfortunately, Trivy continues to give the same warning.

Some less sophisticated scanners will throw false positives like this because the `go.sum` file still lists the old vulnerable version, even though it's not actually used in the binary (due to the pin in the `go.mod` file).

So I next tried using a `replace` directive as this will completely remove the vulnerable version from the `go.sum` file:
```
module github.com/<redacted>

go 1.16

require github.com/fullstorydev/grpcurl v1.8.5

replace golang.org/x/text v0.3.2 => golang.org/x/text v0.3.3 // required to fix CVE-2020-14040
```

Unfortunately the exact same warning continues to be thrown by Trivy.

I double-checked, and there's no mention of `golang.org/text/` `v0.3.2` anywhere in the `go.sum` file or the `go.mod` file. 

1. Where is Trivy getting mention of this? Is it because you have a public DB of `grpcurl` deps and so as soon as it's pulled in you throw a warning even though we're completely overriding it in our source before it gets compiled?
2. Can you fix Trivy to support when `//indirect` pins in `go.mod` override?
3. Can you fix Trivy to support when `replace` directives in `go.mod` override?

**Comment by jeffwidman** [comment]
I also verified the vulnerable version isn't included in the compiled binary... both the `indirect` and `replace` solutions list this:
```
$ go list -m all | grep golang.org/x/text
golang.org/x/text v0.3.3
```

So why does Trivy continue to emit a warning about `v0.3.2` being preset?

**Comment by DmitriyLewen** [comment]
Hello @jeffwidman 
Thanks for your report!
After using `replace golang.org/x/text v0.3.2 => golang.org/x/text v0.3.3` have you cleared the trivy cache?
You can use `trivy --clear-cache` for this.

Regards, Dmitriy

**Comment by jeffwidman** [comment]
Ah, that's it! I had no idea `trivy` had a built-in cache.

So looks like the bug here is the cache wasn't properly cleared for dep bumps implemented via `indirect` / `replace` pins...

**Comment by knqyf263** [comment]
It's weird. The cache should not be used after the binary update. How did you scan it?

**Comment by jeffwidman** [comment]
Saw this exact same error again in a recent build.

Our scan is very vanilla, nothing fancy. We use Drone as our CI engine.

First we build the handful of apps we need into binaries, and then add them to a docker image which is the artifact that we plan to deploy.

Then we scan that docker image using Trivy. We execute the scan from within a custom docker container that another team within our company created using the following Dockerfile:
```
FROM ubuntu:20.04

COPY entrypoint.sh /

RUN apt-get update && apt-get install -y \
        wget \
        apt-transport-https \
        gnupg \
        lsb-release && \
    wget -O - https://aquasecurity.github.io/trivy-repo/deb/public.key | \
        apt-key add - && echo deb https://aquasecurity.github.io/trivy-repo/deb $(lsb_release -sc) main | \
        tee -a /etc/apt/sources.list.d/trivy.list && \
    apt-get update && apt-get install -y \
        trivy && \
    rm -rf /var/lib/apt/lists/* && \
    chmod +x /entrypoint.sh

ENTRYPOINT ["/entrypoint.sh"]
```

The `entrypoint.sh` from that other team isn't what we need, so we instead use our own custom  the following entrypoint commands:
```
# SCAN_IMAGE is the docker image that we want to deploy, think of it more like "image to scan"
$ trivy image --exit-code 0 --severity UNKNOWN,LOW,MEDIUM "$SCAN_IMAGE". 
$ trivy image --exit-code 1 --severity HIGH,CRITICAL "$SCAN_IMAGE"
```

So Trivy goes and scans all the binaries built into the docker image that we plan to deploy. And then if any are `HIGH` or `CRITICAL` it will tell CI to fail the pipeline.

Again, the binary being built into this docker image that Trivy is complaining about already has a pin that I've bumped past the CVE version. As I noted in my comment above, I verified using `go` tools that the vulnerable library is not part of the built binary. So I've no idea why it's flagging that binary for a library version that isn't even part of it.

Plus, having looked at the docs for the trivy cache, it appears to be just the vulnerability database... as best I can tell, it's not caching scan results, so that should be completely unrelated to what's happening here.

If this would be faster to debug via a call, let me know--happy to help.

**Comment by knqyf263** [comment]
We hope it was fixed in v0.26.0.
